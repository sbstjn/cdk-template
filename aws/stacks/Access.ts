import { Stack, StackProps, aws_apigatewayv2 } from 'aws-cdk-lib'
import { IConstruct } from 'constructs'

export interface AccessStackProps extends StackProps {}

export class AccessStack extends Stack {
  api = new aws_apigatewayv2.HttpApi(this, 'api')

  constructor(scope: IConstruct, id: string, props?: AccessStackProps) {
    super(scope, id, props)
  }
}
