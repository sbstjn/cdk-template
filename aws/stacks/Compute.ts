import {
  Duration,
  Stack,
  StackProps,
  aws_iam,
  aws_kms,
  aws_lambda,
  aws_lambda_nodejs,
  aws_sns,
  aws_sns_subscriptions,
  aws_sqs,
} from 'aws-cdk-lib'
import { IConstruct } from 'constructs'

export interface ComputeStackProps extends StackProps {}

export class ComputeStack extends Stack {
  key = new aws_kms.Key(this, 'key', {
    alias: 'compute/key',
    enableKeyRotation: true,
  })

  queue = new aws_sqs.Queue(this, 'queue', {
    visibilityTimeout: Duration.seconds(300),
    encryption: aws_sqs.QueueEncryption.KMS,
    dataKeyReuse: Duration.minutes(5),
    encryptionMasterKey: this.key,
    enforceSSL: true,
  })

  topic = new aws_sns.Topic(this, 'topic', {
    masterKey: this.key,
  })

  publish = new aws_lambda_nodejs.NodejsFunction(this, 'publish', {
    entry: 'src/functions/publish.ts',
    runtime: aws_lambda.Runtime.NODEJS_20_X,
    runtimeManagementMode: aws_lambda.RuntimeManagementMode.AUTO,
    environment: {
      TOPIC_ARN: this.topic.topicArn,
    },
  })

  constructor(scope: IConstruct, id: string, props?: ComputeStackProps) {
    super(scope, id, props)

    this.topic.addSubscription(
      new aws_sns_subscriptions.SqsSubscription(this.queue, {
        rawMessageDelivery: true,
      }),
    )

    this.key.addToResourcePolicy(
      new aws_iam.PolicyStatement({
        sid: 'allow-sns',
        effect: aws_iam.Effect.ALLOW,
        resources: ['*'],
        principals: [new aws_iam.ServicePrincipal('sns'), this.publish.role!],
        actions: ['kms:Decrypt', 'kms:GenerateDataKey'],
      }),
    )

    this.topic.grantPublish(this.publish)
  }
}
