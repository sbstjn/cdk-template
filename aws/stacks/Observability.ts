import { Aspects, Stack, aws_s3 } from 'aws-cdk-lib'
import { Construct, IConstruct } from 'constructs'
import { EnableLambdaXRayTracing } from '../aspects/EnableLambdaXRayTracing'
import { S3BucketAccessLogs } from '../aspects/S3BucketAccessLogs'

export interface ObservabilityStackProps {
  cover: IConstruct[]
}

export class ObservabilityStack extends Stack {
  bucketS3AccessLogs = new aws_s3.Bucket(this, 's3-access-logs')

  constructor(scope: IConstruct, id: string, props: ObservabilityStackProps) {
    super(scope, id)

    props.cover.forEach(node => {
      this.enableS3AccessLogs(node)
      this.enableLambdaXRayTracing(node)
    })
  }

  public enableS3AccessLogs(node: Construct) {
    Aspects.of(node).add(new S3BucketAccessLogs({ bucket: this.bucketS3AccessLogs }))
  }

  public enableLambdaXRayTracing(node: Construct) {
    Aspects.of(node).add(new EnableLambdaXRayTracing())
  }
}
