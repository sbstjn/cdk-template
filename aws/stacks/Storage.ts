import { Stack, StackProps, aws_kms, aws_s3 } from 'aws-cdk-lib'
import { IConstruct } from 'constructs'

export interface StorageStackProps extends StackProps {}

export class StorageStack extends Stack {
  key = new aws_kms.Key(this, 'key', {
    alias: 'storage/key',
    enableKeyRotation: true,
  })

  constructor(scope: IConstruct, id: string, props?: StorageStackProps) {
    super(scope, id, props)

    const bucketProps = {
      bucketKeyEnabled: true,
      encryptionKey: this.key,
      enforceSSL: true,
      blockPublicAccess: aws_s3.BlockPublicAccess.BLOCK_ALL,
    }

    new aws_s3.Bucket(this, 'bucket-one', {
      ...bucketProps,
    })

    new aws_s3.Bucket(this, 'bucket-two', {
      ...bucketProps,
    })
  }
}
