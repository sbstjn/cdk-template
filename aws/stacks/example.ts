import * as sns from '@aws-cdk/aws-sns'
import * as subs from '@aws-cdk/aws-sns-subscriptions'
import * as lambda from '@aws-cdk/aws-lambda-nodejs'
import * as sqs from '@aws-cdk/aws-sqs'
import * as iam from '@aws-cdk/aws-iam'
import * as kms from '@aws-cdk/aws-kms'
import * as cdk from '@aws-cdk/core'

export class ExampleStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props)

    const key = new kms.Key(this, 'YourKey')
    const keyAlias = key.addAlias('your/alias')

    const queue = new sqs.Queue(this, 'YourQueue', {
      visibilityTimeout: cdk.Duration.seconds(300),
      encryption: sqs.QueueEncryption.KMS,
      dataKeyReuse: cdk.Duration.minutes(5),
      encryptionMasterKey: keyAlias,
    })

    const topic = new sns.Topic(this, 'YourTopic', {
      masterKey: keyAlias,
    })

    topic.addSubscription(
      new subs.SqsSubscription(queue, {
        rawMessageDelivery: true,
      })
    )

    const func = new lambda.NodejsFunction(this, "YourLambda", {
      entry: "src/example/handler.ts",
      handler: "run",
      environment: {
        TOPIC_ARN: topic.topicArn,
      },
    })

    keyAlias.addToResourcePolicy(new iam.PolicyStatement({
      sid: 'sns-allow',
      effect: iam.Effect.ALLOW,
      resources: ['*'],
      principals: [
        new iam.ServicePrincipal('sns'),
        func.role!
      ],
      actions: [
        'kms:Decrypt',
        'kms:GenerateDataKey'
      ]
    }))

    topic.grantPublish(func)
  }
}
