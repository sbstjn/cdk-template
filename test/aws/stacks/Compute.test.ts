import { App, Stack } from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'

import { ComputeStack } from '../../../aws/stacks/Compute'

describe('Stack', () => {
  let app: App
  let stack: Stack
  let template: Template

  beforeAll(() => {
    app = new App({
      context: {
        version: 'v3.2.1-compute',
      },
    })

    stack = new ComputeStack(app, 'compute')
    template = Template.fromStack(stack)
  })

  describe('Template', () => {
    it('match snapshot', () => {
      expect(template.toJSON()).toMatchSnapshot()
    })
  })

  describe('SQS Queue', () => {
    it('is created', () => {
      template.hasResource('AWS::SQS::Queue', {
        Properties: {
          VisibilityTimeout: 300,
        },
      })
    })
  })

  describe('SNS Topic', () => {
    it('is created', () => {
      template.hasResource('AWS::SNS::Topic', {})
    })
  })

  describe('Lambda Function', () => {
    it('is created', () => {
      template.hasResource('AWS::Lambda::Function', {})
    })
  })
})
