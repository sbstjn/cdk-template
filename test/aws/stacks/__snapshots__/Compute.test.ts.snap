// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack Template match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "keyAliasCEB16DF2": {
      "Properties": {
        "AliasName": "alias/compute/key",
        "TargetKeyId": {
          "Fn::GetAtt": [
            "keyFEDD6EC0",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
    "keyFEDD6EC0": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:GenerateDataKey",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:GenerateDataKey",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "publishServiceRoleD6DF777D",
                    "Arn",
                  ],
                },
                "Service": "sns.amazonaws.com",
              },
              "Resource": "*",
              "Sid": "allow-sns",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "publish98FFDE07": {
      "DependsOn": [
        "publishServiceRoleDefaultPolicy1754110B",
        "publishServiceRoleD6DF777D",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "1cda00555f4d343822e292e6f49c4910db8a409e4fdb0bf7c460d72029bb2c28.zip",
        },
        "Environment": {
          "Variables": {
            "TOPIC_ARN": {
              "Ref": "topic69831491",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "publishServiceRoleD6DF777D",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "RuntimeManagementConfig": {
          "UpdateRuntimeOn": "Auto",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "publishServiceRoleD6DF777D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "publishServiceRoleDefaultPolicy1754110B": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "topic69831491",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "publishServiceRoleDefaultPolicy1754110B",
        "Roles": [
          {
            "Ref": "publishServiceRoleD6DF777D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "queue276F7297": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsDataKeyReusePeriodSeconds": 300,
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "keyFEDD6EC0",
            "Arn",
          ],
        },
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "queuePolicy89DB7105": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "queue276F7297",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Ref": "topic69831491",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "queue276F7297",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "queue276F7297",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "queuecomputetopic73AC0DBD00649C29": {
      "DependsOn": [
        "queuePolicy89DB7105",
      ],
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "queue276F7297",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "RawMessageDelivery": true,
        "TopicArn": {
          "Ref": "topic69831491",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "topic69831491": {
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "keyFEDD6EC0",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
